# <span style="color: Orange;"> Credit Card Transaction Analysis - PART 1: Credit  card company POV </span>

### *<span style= "color:gray;"> Thought process, Insights and Recommendations</span>*

### *Context* :  
 __Bank "ABC" had inaugurated offering their credit card facility on 1st Oct, 2013.__  
__They started collecting data on the credit card transaction details since then.__  
    
__Now, after nearly two years, they have a stakeholder meeting on June 1st 2015 to assess the situation regarding the growth in the credit card section.__  

__The data till now (latest data being 26th May 2015) needs to be analysed and <span style="color:green;">insights and recommendations needs to be presented to the stakeholders during their meeting so that they can take data-driven decision on how to proceed.__</span>

### Source of data:  
*Credit_card_transactions_India.csv*

### *<span style= "color:gray;"> STEP 1: Getting to know the data</span>*

<span style="font-size:14px;"> __Let's start with getting the look and feel of the data that we need to work with.__ 
    
__We would be looking at the number of rows and columns of the data along with what each column denotes and the data types.__ 
    
__We would also be noting various obvious issues/ anomalies with the data that needs to be fixed as a part of cleaning the data although data cleaning is not the mail goal of this section.__ </span>

In [53]:
import numpy as np
import pandas as pd

In [54]:
original_df = pd.read_csv(r"D:\Data Analysis\Datasets\Credit_card_transactions_India.csv")

In [55]:
working_df = original_df.copy(deep=True)
working_df

index                   City       Date  Card Type Exp Type Gender  \
0          0           Delhi, India  29-Oct-14       Gold    Bills      F   
1          1  Greater Mumbai, India  22-Aug-14   Platinum    Bills      F   
2          2       Bengaluru, India  27-Aug-14     Silver    Bills      F   
3          3  Greater Mumbai, India  12-Apr-14  Signature    Bills      F   
4          4       Bengaluru, India   5-May-15       Gold    Bills      F   
...      ...                    ...        ...        ...      ...    ...   
26047  26047         Kolkata, India  22-Jun-14     Silver   Travel      F   
26048  26048            Pune, India   3-Aug-14  Signature   Travel      M   
26049  26049       Hyderabad, India  16-Jan-15     Silver   Travel      M   
26050  26050          Kanpur, India  14-Sep-14     Silver   Travel      M   
26051  26051       Hyderabad, India  19-Oct-13  Signature    Bills      M   

       Amount  
0       82475  
1       32555  
2      101738  
3      123424  
4      171574  
...       ...  
26047  128191  
26048  246316  
26049  265019  
26050   88174  
26051  184410  

[26052 rows x 7 columns]


- __There are 6 columns (excluding index) and 26k rows__ 
- __First things first, the index column provided in the file needs to be dropped.__ 
- __The city column has the format: "city name", India.  
    We can drop the comma and India, as it doesn't add any value to the analysis and just retain the actual city name.__ 
- __The dates don't seem to be in chronological order, so the dataframe has to be sorted with dates in ascending order.__
    



    
### <span style= "color:gray;"> Column_name: Description (expected data type ): </span>

*City: The city in which the transaction took place. (String)* 

*Date: The date of the transaction. (Date)*  

*Card Type: The type of credit card used for the transaction. (String)*  

*Exp Type: The type of expense associated with the transaction. (String)*  

*Gender: The gender of the cardholder. (String)*  

*Amount: The amount of the transaction. (Number)*

__Before we do any further analysis, let's correct the obvious issues__

###  <span style="color:red;">Let's start by dropping the in-built index column </span>

In [56]:
# Dropping in-built Index column
working_df = working_df.drop('index', axis=1)
working_df

City       Date  Card Type Exp Type Gender  Amount
0               Delhi, India  29-Oct-14       Gold    Bills      F   82475
1      Greater Mumbai, India  22-Aug-14   Platinum    Bills      F   32555
2           Bengaluru, India  27-Aug-14     Silver    Bills      F  101738
3      Greater Mumbai, India  12-Apr-14  Signature    Bills      F  123424
4           Bengaluru, India   5-May-15       Gold    Bills      F  171574
...                      ...        ...        ...      ...    ...     ...
26047         Kolkata, India  22-Jun-14     Silver   Travel      F  128191
26048            Pune, India   3-Aug-14  Signature   Travel      M  246316
26049       Hyderabad, India  16-Jan-15     Silver   Travel      M  265019
26050          Kanpur, India  14-Sep-14     Silver   Travel      M   88174
26051       Hyderabad, India  19-Oct-13  Signature    Bills      M  184410

[26052 rows x 6 columns]

In [57]:
#Sorting chronologically
#working_df['Date'] = pd.to_datetime(working_df['Date'])
working_df = working_df.sort_values(by='Date', ascending=True)
working_df

City      Date  Card Type       Exp Type Gender  Amount
24582       Chennai, India  1-Apr-14   Platinum  Entertainment      F   93038
16742  Wadgaon Road, India  1-Apr-14       Gold           Food      M  274661
4305      Bengaluru, India  1-Apr-14   Platinum          Bills      F  174903
20368        Jaipur, India  1-Apr-14   Platinum  Entertainment      M  118570
15309         Nokha, India  1-Apr-14  Signature          Bills      M  163677
...                    ...       ...        ...            ...    ...     ...
12607     Bengaluru, India  9-Sep-14       Gold           Food      M  259100
3136          Delhi, India  9-Sep-14   Platinum           Food      F  217217
17742         Udupi, India  9-Sep-14       Gold  Entertainment      M  105855
7435          Delhi, India  9-Sep-14  Signature           Food      M   46182
15194  Rajagangapur, India  9-Sep-14  Signature           Food      M  165047

[26052 rows x 6 columns]

###  <span style="color:maroon;">The above result shows the importance of making sure the date column has the correct data type </span>  

Let's check the data types of these columns

In [58]:
working_df.dtypes

City         object
Date         object
Card Type    object
Exp Type     object
Gender       object
Amount        int64
dtype: object

 #### <span style="color:red;">  data type of "Date" column is object. </span>  
 
 ### <span style="color:red;"> Let's change it to datetime and see how the new sorting is</span>  

In [59]:
# changing date column data type

working_df['Date'] = pd.to_datetime(working_df['Date'])
working_df.dtypes

City                 object
Date         datetime64[ns]
Card Type            object
Exp Type             object
Gender               object
Amount                int64
dtype: object

In [60]:
#Sorting chronologically after changing date column data type

working_df = working_df.sort_values(by='Date', ascending=True)
working_df

City       Date  Card Type       Exp Type Gender  \
72              Delhi, India 2013-10-04  Signature          Bills      F   
7201   Greater Mumbai, India 2013-10-04       Gold  Entertainment      M   
6931        Bengaluru, India 2013-10-04       Gold           Fuel      F   
8645        Bengaluru, India 2013-10-04     Silver          Bills      M   
23691         Chennai, India 2013-10-04       Gold        Grocery      M   
...                      ...        ...        ...            ...    ...   
7891        Ahmedabad, India 2015-05-26   Platinum        Grocery      F   
24604          Jaipur, India 2015-05-26     Silver           Fuel      F   
9980            Delhi, India 2015-05-26   Platinum  Entertainment      F   
22648         Lucknow, India 2015-05-26       Gold        Grocery      M   
14658          Washim, India 2015-05-26  Signature           Fuel      F   

       Amount  
72     550782  
7201   201942  
6931   236037  
8645   159769  
23691   27511  
...       ...  
7891   269998  
24604  292561  
9980   200932  
22648  249011  
14658  223068  

[26052 rows x 6 columns]

__Glancing at the datframe output:__

__Oldest entry (first entry) is on 4th Oct 2013 (2013-10-04) and__ 

__Latest entry (last entry) is on 26th May 2015 (2015-05-26)__

### <span style="color:red;"> Let's check the oldest and latest entry to confirm that the sorting is correct. </span>

In [61]:
# checking the oldest entry and the latest entry
#dt.date extracts just the date part of the datetime value

oldest_entry = working_df["Date"].dt.date.min() 
latest_entry = working_df["Date"].dt.date.max()
print("oldest entry: ", oldest_entry)
print("latest entry: ", latest_entry)

oldest entry:  2013-10-04
latest entry:  2015-05-26


__sorting seems to have worked properly as the oldest and latest values cross-checks with the dates visible in the dataframe output__

 ### <span style="color:red;">  but now the index is all jumbled, so let's reset it. </span> 
Note that on resetting, the current index would be added as a new column, hence we need to set drop=True for dropping the older, jumbled index

In [62]:
# Resetting index
working_df = working_df.reset_index(drop=True)
working_df

City       Date  Card Type       Exp Type Gender  \
0               Delhi, India 2013-10-04  Signature          Bills      F   
1      Greater Mumbai, India 2013-10-04       Gold  Entertainment      M   
2           Bengaluru, India 2013-10-04       Gold           Fuel      F   
3           Bengaluru, India 2013-10-04     Silver          Bills      M   
4             Chennai, India 2013-10-04       Gold        Grocery      M   
...                      ...        ...        ...            ...    ...   
26047       Ahmedabad, India 2015-05-26   Platinum        Grocery      F   
26048          Jaipur, India 2015-05-26     Silver           Fuel      F   
26049           Delhi, India 2015-05-26   Platinum  Entertainment      F   
26050         Lucknow, India 2015-05-26       Gold        Grocery      M   
26051          Washim, India 2015-05-26  Signature           Fuel      F   

       Amount  
0      550782  
1      201942  
2      236037  
3      159769  
4       27511  
...       ...  
26047  269998  
26048  292561  
26049  200932  
26050  249011  
26051  223068  

[26052 rows x 6 columns]

 ### <span style="color:red;"> Now let's split the City column. </span>

In [63]:
working_df[['City', 'Country']] = working_df['City'].str.split(',', expand=True)
working_df

City       Date  Card Type       Exp Type Gender  Amount  \
0               Delhi 2013-10-04  Signature          Bills      F  550782   
1      Greater Mumbai 2013-10-04       Gold  Entertainment      M  201942   
2           Bengaluru 2013-10-04       Gold           Fuel      F  236037   
3           Bengaluru 2013-10-04     Silver          Bills      M  159769   
4             Chennai 2013-10-04       Gold        Grocery      M   27511   
...               ...        ...        ...            ...    ...     ...   
26047       Ahmedabad 2015-05-26   Platinum        Grocery      F  269998   
26048          Jaipur 2015-05-26     Silver           Fuel      F  292561   
26049           Delhi 2015-05-26   Platinum  Entertainment      F  200932   
26050         Lucknow 2015-05-26       Gold        Grocery      M  249011   
26051          Washim 2015-05-26  Signature           Fuel      F  223068   

      Country  
0       India  
1       India  
2       India  
3       India  
4       India  
...       ...  
26047   India  
26048   India  
26049   India  
26050   India  
26051   India  

[26052 rows x 7 columns]

 ### <span style="color:red;"> We can remove the country column as it adds no value for analysis. </span>
 
 #### <span style="color:red;">But wait, are we sure it only contains India? </span> 
 
 Let's check the unique values in that column to confirm before dropping

In [64]:
# Checking the unique entries in "Country" column

working_df['Country'].unique()

array([' India'], dtype=object)

__Looks good as it only seems to contain India.__

### <span style="color:red;"> Let's drop "Country" column. </span>

In [65]:
# dropping country column

working_df = working_df.drop('Country', axis=1)
working_df

City       Date  Card Type       Exp Type Gender  Amount
0               Delhi 2013-10-04  Signature          Bills      F  550782
1      Greater Mumbai 2013-10-04       Gold  Entertainment      M  201942
2           Bengaluru 2013-10-04       Gold           Fuel      F  236037
3           Bengaluru 2013-10-04     Silver          Bills      M  159769
4             Chennai 2013-10-04       Gold        Grocery      M   27511
...               ...        ...        ...            ...    ...     ...
26047       Ahmedabad 2015-05-26   Platinum        Grocery      F  269998
26048          Jaipur 2015-05-26     Silver           Fuel      F  292561
26049           Delhi 2015-05-26   Platinum  Entertainment      F  200932
26050         Lucknow 2015-05-26       Gold        Grocery      M  249011
26051          Washim 2015-05-26  Signature           Fuel      F  223068

[26052 rows x 6 columns]

### Now we are ready for some more EDA

### <span style="color:red;"> Let's check how many unique cities are there. </span>

In [66]:
# Number of unique cities and their names

print('Number of unique cities: ', len(working_df['City'].unique()))

Number of unique cities:  986


### <span style="color:purple;"> printing out all 986 names wouldn't help much, hence let's skip that step. </span>

### <span style="color:red;"> Let's check how many unique card types are there. </span>

In [67]:
# Number of unique card types and their names

print('Number of unique card types: ', len(working_df['Card Type'].unique()))

Number of unique card types:  4


In [68]:
print('Name of unique card types: \n', working_df['Card Type'].unique())

Name of unique card types: 
 ['Signature' 'Gold' 'Silver' 'Platinum']


### <span style="color:purple;"> The possible hierarchy would be: </span>  

- __Signature__  

- __Platinum__  

- __Gold__  

- __Silver__

*Assumed from the fact that Visa is the most notable providers of Signature card type which is above their Platinum*